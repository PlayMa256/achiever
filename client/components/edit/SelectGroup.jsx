import React from 'react';
import PropTypes from 'prop-types';

const renderOptions = options =>
	options.map(option => (
		<option key={option.id} value={option.id}>
			{option.name}
		</option>
	));

const SelectGroup = ({
	name,
	label,
	options,
	selected,
	onChange
}) => (
	<div className="select-group">
		<label htmlFor={name}>{label}</label>
		<select
			name={name}
			className="detail-selector"
			value={selected || ''}
			onChange={event => onChange(event.target.value)}
		>
			{ renderOptions(options, selected) }
		</select>
	</div>
);

SelectGroup.propTypes = {
	name: PropTypes.string.isRequired,
	label: PropTypes.string.isRequired,
	options: PropTypes.arrayOf(PropTypes.object),
	selected: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
	onChange: PropTypes.func
};

SelectGroup.defaultProps = {
	selected: '',
	options: [],
	onChange: null
};

export default SelectGroup;
