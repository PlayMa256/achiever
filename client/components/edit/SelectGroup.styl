@import "_colours.styl"
@import "_standards.styl"

.select-group
	min-height: 40px;
	label
		label-style();
	select
		box-style();
		width: 230px;
		font-size: 13px;
		padding: 4px;
		background-color: white;
		margin-bottom: 20px;
	.select-group-replacement
		font-size: 14px;