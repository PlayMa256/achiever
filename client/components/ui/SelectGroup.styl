@import "../../styles/_colours.styl"
@import "../../styles/_standards.styl"

.SelectGroup
	min-height: 40px;
	.label
		width 60px
		margin-right 10px
		font-size 11px
		float left
		text-align right
		clear: both;
		height: 30px;
		display: flex;
		flex-direction: column;
		justify-content: center;
	select
		box-style();
		width: 230px;
		font-size: 13px;
		padding: 4px;
		background-color: white;
		margin-bottom: 20px;
	div
		font-size: 14px;
		height: 30px;
		display: flex;
		flex-direction: column;
		justify-content: center;
