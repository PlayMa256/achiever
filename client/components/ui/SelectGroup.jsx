import React from 'react';
import PropTypes from 'prop-types';

import './SelectGroup.styl';

const renderOptions = options => options.map(option => (
	<option key={option.id} value={option.id}>
		{option.name}
	</option>
));

const SelectGroup = ({
	label,
	options,
	selected,
	onChange,
	showTextInstead,
	isDisabled
}) => (
	<fieldset className="SelectGroup">
		<label>
			<span className="label">{ label }</span>
			{ showTextInstead ?
				<div className="select-group-replacement">
					{ showTextInstead }
				</div> :
				<select
					className="detail-selector"
					value={selected || ''}
					onChange={event => onChange(event.target.value)}
					disabled={isDisabled}
				>
					{ renderOptions(options, selected) }
				</select>
			}
		</label>
	</fieldset>
);

SelectGroup.propTypes = {
	label: PropTypes.string.isRequired,
	options: PropTypes.arrayOf(PropTypes.shape({
		id: PropTypes.number,
		name: PropTypes.string
	})),
	selected: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
	onChange: PropTypes.func,
	showTextInstead: PropTypes.string,
	isDisabled: PropTypes.bool
};

SelectGroup.defaultProps = {
	selected: '',
	options: [],
	onChange: null,
	showTextInstead: null,
	isDisabled: false
};

export default SelectGroup;
