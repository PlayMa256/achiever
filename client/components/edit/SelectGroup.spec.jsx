import React from 'react';
import { shallow } from 'enzyme';

import SelectGroup from './SelectGroup';

const mockOptions = [
	{ id: 1, name: 'one' },
	{ id: 2, name: 'two' },
	{ id: 3, name: 'three' }
];

describe('SelectGroup', () => {
	describe('render', () => {
		it('should show a select component with min props', () => {
			const wrapper = shallow(<SelectGroup name="testName" label="testLabel" />);
			expect(wrapper).toMatchSnapshot();
		});
		it('should call onChange with selected option', () => {
			const onChangeFn = jest.fn();
			const wrapper = shallow(<SelectGroup
				name="testName"
				label="testLabel"
				options={mockOptions}
				onChange={onChangeFn}
			/>);
			const selectField = wrapper.find('select');
			selectField.simulate('change', { target: { value: 2 } });
			expect(onChangeFn).toHaveBeenCalledWith(2);
		});
	});
});
