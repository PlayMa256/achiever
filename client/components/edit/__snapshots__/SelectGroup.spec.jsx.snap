// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectGroup render should show a select component with min props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectGroup
    disabled={false}
    label="testLabel"
    name="testName"
    onChange={null}
    options={Array []}
    selected=""
    showTextInstead={null}
    tabIndex={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          htmlFor="testName"
        >
          testLabel
        </label>,
        <select
          className="detail-selector"
          disabled={false}
          name="testName"
          onChange={[Function]}
          tabIndex={0}
          value=""
        />,
      ],
      "className": "select-group",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "testLabel",
          "htmlFor": "testName",
        },
        "ref": null,
        "rendered": "testLabel",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "detail-selector",
          "disabled": false,
          "name": "testName",
          "onChange": [Function],
          "tabIndex": 0,
          "value": "",
        },
        "ref": null,
        "rendered": Array [],
        "type": "select",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="testName"
          >
            testLabel
          </label>,
          <select
            className="detail-selector"
            disabled={false}
            name="testName"
            onChange={[Function]}
            tabIndex={0}
            value=""
          />,
        ],
        "className": "select-group",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "testLabel",
            "htmlFor": "testName",
          },
          "ref": null,
          "rendered": "testLabel",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "detail-selector",
            "disabled": false,
            "name": "testName",
            "onChange": [Function],
            "tabIndex": 0,
            "value": "",
          },
          "ref": null,
          "rendered": Array [],
          "type": "select",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
